drivers/store/redis/store.go

```diff
@@ -12,24 +14,60 @@ import (
 	"github.com/ulule/limiter/v3/drivers/store/common"
 )
 
+const (
+	luaIncrScript = `
+local key = KEYS[1]
+local count = tonumber(ARGV[1])
+local ttl = tonumber(ARGV[2])
+local ret = redis.call("incrby", key, ARGV[1])
+if ret == count then
+	if ttl > 0 then
+		redis.call("pexpire", key, ARGV[2])
+	end
+	return {ret, ttl}
+end
+ttl = redis.call("pttl", key)
+return {ret, ttl}
+`
+	luaPeekScript = `
+local key = KEYS[1]
+local v = redis.call("get", key)
+if v == false then
+	return {0, 0}
+end
+local ttl = redis.call("pttl", key)
+return {tonumber(v), ttl}
+`
+)
+
 // Client is an interface thats allows to use a redis cluster or a redis single client seamlessly.
 type Client interface {
 	Get(ctx context.Context, key string) *libredis.StringCmd
 	Set(ctx context.Context, key string, value interface{}, expiration time.Duration) *libredis.StatusCmd
 	Watch(ctx context.Context, handler func(*libredis.Tx) error, keys ...string) error
 	Del(ctx context.Context, keys ...string) *libredis.IntCmd
 	SetNX(ctx context.Context, key string, value interface{}, expiration time.Duration) *libredis.BoolCmd
-	Eval(ctx context.Context, script string, keys []string, args ...interface{}) *libredis.Cmd
+	EvalSha(ctx context.Context, sha string, keys []string, args ...interface{}) *libredis.Cmd
+	ScriptLoad(ctx context.Context, script string) *libredis.StringCmd
 }
 
 // Store is the redis store.
 type Store struct {
 	// Prefix used for the key.
 	Prefix string
+	// deprecated, this option make no sense when all operations were atomic
 	// MaxRetry is the maximum number of retry under race conditions.
 	MaxRetry int
 	// client used to communicate with redis server.
 	client Client
+	// luaIncrSHA is the SHA of increase and expire key script
+	luaIncrSHA string
+	// luaPeekSHA is the SHA of peek and expire key script
+	luaPeekSHA string
+	// hasLuaScriptLoaded was used to check whether the lua script was loaded or not
+	hasLuaScriptLoaded bool
+	// mu was used to avoid multiple script load in a context of high concurrency
+	mu sync.Mutex
```

Comment:
Mutex sould be before the fields that they "protect" against race condition.

For a simple boolean, you could use [Compare and Swap](https://en.wikipedia.org/wiki/Compare-and-swap) primitive that are more efficient than an mutex.

You could use [atomic.CompareAndSwapUint32](https://golang.org/pkg/sync/atomic/#CompareAndSwapUint32) for that.

Reply:
Thanks for your excellent advice, I have changed it to CAS instead of the mutex, You can have a look again.

: )

Reply:
It seems has a corner case if we are using CAS, other goroutines may retry evalSHA again before loading scripts if we use CAS before scripts loading, and may cause multi scripts loaded if uses CAS after scripts loading. BTW, thanks for your advice and review. : )

Reply:
By curiosity, could you tell me how you have setup this tests, and detect it ? 
Are you using the benchmark or do you it on real instance ?
Thank you.

Reply:
Just replace the redis/store.go with the Lua script version